gtda/homology/simplicial.py

```diff
@@ -721,3 +722,272 @@ def plot(Xt, sample=0, homology_dimensions=None):
         """
         return plot_diagram(
             Xt[sample], homology_dimensions=homology_dimensions)
+
+
+@adapt_fit_transform_docs
+class FlagserPersistence(BaseEstimator, TransformerMixin, PlotterMixin):
+    """:ref:`Persistence diagrams <persistence_diagram>` resulting from
+    :ref:`filtrations <filtered_complex>` of :ref:`directed or undirected flag
+    complexes <clique_and_flag_complexes>`.
+
+    Given a weighted directed or undirected graph, information about the
+    appearance and disappearance of topological features (technically,
+    :ref:`homology classes <homology_and_cohomology>`) of various dimension
+    and at different scales is summarised in the corresponding persistence
+    diagram.
+
+    Parameters
+    ----------
+    homology_dimensions : list or tuple, optional, default: ``(0, 1)``
+        Dimensions (non-negative integers) of the topological features to be
+        detected.
+
+    directed : bool, optional, default: ``True``
+        If ``True``, :meth:`transform` computes the persistence diagrams of
```

Comment:
I'd like to suggest a rewording of this section as follows:
```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected. When
        ``False``, it is sufficient to pass a collection of upper-triangular
        matrices as input to :meth:`transform` because, if two directed edges
        corresponding to the same undirected edge are assigned different
        weights, only the one in the upper triangular part of the adjacency
        matrix is considered.
```

Reply:
The logic of the last sentence is not clear to me:

>        When ``False``, it is sufficient to pass a collection of upper-triangular
>         matrices as input to :meth:`transform` because, if two directed edges
>         corresponding to the same undirected edge are assigned different
>         weights, only the one in the upper triangular part of the adjacency
>         matrix is considered.

It is sufficient because the weight on only one of the oriented edge is enough to assign a weight to the unoriented edge. The last part of the sentence deals with the edge case where two oriented edges corresponding to the same unoriented edge have different weights.

Reply:
I still see it slightly differently:
- If the input is dense, there are always two explicit weights per undirected edge;
- If the input is sparse, there are still two weights per undirected edge because we say in the sparse case that absence of edge is actually meaningful.

Reply:
If we referred to the docstrings for transform here then my second bullet point would be more compelling I think.

Reply:
I think the reality is more complicated:

- If the input is dense and not symmetric but some values of the upper-trianguar part are larger than `max_edge_weight`, the value on the lower-trinagular part will be considered. This is also true in the sparse case actually.
- If the inpur is sparse, and there is no entry on an edge in the upper-triangular part, the value of the unoriented edge will be using the weight in the lower-triangular part.

Reply:
Thanks for the feedback. What's the reason for the behaviour in your first bullet point? It seems a bit confusing to me.

Reply:
If the weight is higher than `max_edge_weight`, it is absent from the filtration and therefore not passed to C++ `flagser`.

Reply:
Sure, but why is the lower weight picked instead? In my opinion, the first thing we should do in the dense case is check for disagreements between off-diagonal entries, and resolve them in favour of then upper diagonal. Then the issue you describe would not arise.

Reply:
But this is what we have for now so we are stuck with documenting this weird (from the user perspective) behaviour. I'll try to come up with an alternative!

Reply:
```        
        Which death value to assign to features which are still alive at
        filtration value `max_edge_length`. ``None`` means that this
        death value is declared to be equal to `max_edge_length`.
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected, and:

        - if `max_edge_weight` is `numpy.inf`, it is sufficient to pass a
          collection of (dense or sparse) upper-triangular matrices;
        - if `max_edge_weight` is finite, it is recommended to pass either a
          collection of symmetric dense matrices, or a collection of sparse
          upper-triangular matrices;
        - if two directed edges corresponding to the same undirected edge are
          explicitly assigned different weights and neither exceeds
          `max_edge_weight`, only the one in the upper triangular part of the
          adjacency matrix is considered.
```

What do you think? If you agree, it might be a good idea to reference to this again in the docstrings for `fit` and `transform`.

Reply:
I feel that with something like the above we would be guiding the user a little more directly.

Reply:
As I read this, it definitely sounds too complex to be fully explicitly documented. I would suggest not explaining all those rare edge cases but rather explain the user what type of adjacency matrix has to be passed and why. 

```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected. In
        that case, if two directed edges corresponding to the same 
        undirected edge are explicitly assigned different weights and neither 
        exceeds `max_edge_weight`, only the one in the upper triangular part 
        of the adjacency matrix is considered. Therefore, it is recommended to 
        pass either a collection of symmetric or upper-triangular matrices.
```

An alternative would also be to force a simpler behavior by systematically enforcing that the lower triangular part of the matrix is equal to the upper triangular part.

Reply:
Thanks for the feedback. I agree about being prescriptive rather than comprehensively descriptive, but in that case I would take a different route and avoid altogether to discuss what happens when different weights are assigned. I'd simply tell the user which behaviour is safe. Simply removing the final point in the suggestion I had above would do the job:
```
        If ``True``, :meth:`transform` computes the persistence diagrams of
        the filtered directed flag complexes arising from the input collection
        of weighted directed graphs. If ``False``, :meth:`transform` computes
        the persistence diagrams of the filtered undirected flag complexes
        obtained by regarding all input weighted graphs as undirected, and:

        - if `max_edge_weight` is ``numpy.inf``, it is sufficient to pass a
          collection of (dense or sparse) upper-triangular matrices;
        - if `max_edge_weight` is finite, it is recommended to pass either a
          collection of symmetric dense matrices, or a collection of sparse
          upper-triangular matrices.
```

Sorry to be a pain! I do think we have come a long way since the first versions though :)

Reply:
Agreed, that works.